pkg/agent/openflow/packetin.go

```diff
@@ -27,51 +27,81 @@ type PacketInHandler interface {
 }
 
 const (
-	// Action explicitly output to controller.
-	ofprAction ofpPacketInReason = 1
 	// Max packetInQueue size.
 	packetInQueueSize int = 256
 )
 
-func (c *client) RegisterPacketInHandler(packetHandlerName string, packetInHandler interface{}) {
+func NewOFReason(value uint8) ofpPacketInReason{
+	return ofpPacketInReason(value)
+}
+
+func (c *client) RegisterPacketInHandler(packetHandlerReason ofpPacketInReason, packetHandlerName string, packetInHandler interface{}) {
 	handler, ok := packetInHandler.(PacketInHandler)
 	if !ok {
-		klog.Errorf("Invalid Traceflow controller.")
+		klog.Errorf("Invalid controller.")
 		return
 	}
-	c.packetInHandlers[packetHandlerName] = handler
+	if c.packetInHandlers[packetHandlerReason] == nil {
+		c.packetInHandlers[packetHandlerReason] = map[string]PacketInHandler{}
+	}
+	c.packetInHandlers[packetHandlerReason][packetHandlerName] = handler
 }
 
-func (c *client) StartPacketInHandler(stopCh <-chan struct{}) {
-	if len(c.packetInHandlers) == 0 {
+func (c *client) StartPacketInHandler(packetInStartedReason []uint8, stopCh <-chan struct{}) {
+	if len(c.packetInHandlers) == 0 || len(packetInStartedReason) == 0 {
 		return
 	}
-	ch := make(chan *ofctrl.PacketIn)
-	err := c.SubscribePacketIn(uint8(ofprAction), ch)
+	// Subscribe packetin for TraceFlow with reason[0], using reason 1 in ovs
+	tfCh := make(chan *ofctrl.PacketIn)
+	err := c.SubscribePacketIn(packetInStartedReason[0], tfCh)
```

Comment:
This function is always called with both the reasons, even though one of the feature gate is disabled (Traceflow or AntreaPolicy).  What happens with the channel that has been created? Is the expectation that we define the resources for that channel and never use it?

Reply:
wondering if we should do another check here for each enabled feature gate?

Reply:
It is possible to create channels based on passed in feature gates, however in that case, I did not find a proper solution to select dynamically (such as for a channel array). Any suggestion?

Reply:
Is it better to add an internal object to maintain: reason, channel, workqueue, and the logic of processing packetIn messages. Then every time a new feature registers the handler, we could create such an obejct. Each object could have its own goroutine to run the logics. And  we don't need to bind all features together.

Reply:
Thanks for the suggestion, does that mean having separate goroutines to select between the object's channel or stop channel?

Reply:
If you use a seperate goroutine for every feature, I think it could only focus on its own packetIn message channel and the stop channel, and ignore the messages for other features.

Reply:
Created an interface for each feature, so now traceflow and networkpolicy are of different objects. Update in openflow/packetin.go. Let me know of any feedback. Thanks!